### PR TITLE
[FOLLOWUP][SPARK-36967][CORE] Report accurate shuffle block size if its skewed

### PR DESCRIPTION
**Pull request by wankunde**
### What changes were proposed in this pull request?

Now we will sort the shuffle blocks twice in map side to find out the skewed blocks. 
We can add a flag in `Utils.median(sizes: Array[Long])` to avoid sort the input array.

### Why are the changes needed?

To avoid additional sort.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing UTs

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -3221,11 +3221,12 @@ private[spark] object Utils extends Logging {
    * Return the median number of a long array
    *
    * @param sizes
+   * @param alreadySorted
    * @return
    */
-  def median(sizes: Array[Long]): Long = {
+  def median(sizes: Array[Long], alreadySorted: Boolean = false): Long = {
```
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:3227]: What about always assuming the `sizes` are sorted (mentioning this in the description and reflecting in the parameter name ie. `sortedSizes` or in the name of the method..)?

Comment (wankunde) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3227]: Hi, @attilapiros , in my opinion,  `median` method should support unsorted input array. But if we change the parameter name to `sortedSizes`, that will be a little confusing?

Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:3227]: Ok, but in that case I won't give a default value for the `alreadySorted`, let's require the client to define it.

Comment (wankunde) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: Agree, I have updated `median` method.

File: core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala [hunk 2]
```diff
@@ -265,7 +265,7 @@ private[spark] object HighlyCompressedMapStatus {
     val threshold =
       if (accurateBlockSkewedFactor > 0) {
         val sortedSizes = uncompressedSizes.sorted
-        val medianSize: Long = Utils.median(sortedSizes)
+        val medianSize: Long = Utils.median(sortedSizes, true)
```
Comment (zhengruifeng) [inline core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala:268]: only in this call site `alreadySorted` is `true`, what about only changing this place to utilize the order?

Comment (wankunde) [reply core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala:268]: Hi, @zhengruifeng Originally I wrote the median function code here, but  @JoshRosen suggested extracting the common code into a util function https://github.com/apache/spark/pull/34234#discussion_r731369282 .
Either changing the median method or copying some code here all makes sense to me.
WDYT?

Comment (zhengruifeng) [reply core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala:268]: ok, I am netural on it.

**Comment by ulysses-you** [comment]
can we go ahead since 3.3.0 will be cut soon. thank you ! also cc @mridulm @Ngone51

**Review by attilapiros — APPROVED** [review]
LGTM

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
Merged to master

**Comment by ulysses-you** [comment]
thank you all !

**Comment by mridulm** [comment]
late LGTM, thanks for adding this @wankunde !

**Pull request closed without merging**